### PR TITLE
Refactor Mixer Adapter CRDs

### PR DESCRIPTION
**Pull request by douglas-reid**
Mixer supports a model of individual CRDs for each adapter type. This means that, at startup time, Mixer must establish watches on a large number of CRs. This can add load to the API Server, especially in HPA scenarios. This can contribute to issues such as #7675 and #8184 .

This PR adds support in Mixer for static handler configuration using the `handler` CRD. This will enable Mixer to transition off of the CRD-per-adapter model, hopefully significantly reducing the config burden and harmonizing the config story across compiled-in adapters and out-of-process adapters.

This PR does three main things:
1. Adds a `UseAdapterCRDs` flag to Mixer (defaults to `true` at the moment for backwards compatibility). This will enable deployments to turn off a large chunk of the required config watches.
1. Updates the install artifacts to use the new-style handler configs. (`mixer/testdata/config` resources are updated as well for consistency).
1. Updates the `emphemeral.go` code to create static handlers with names of `name + '.' + namespace` from the `handler` CRs.

### CODE REVIEW

File: install/kubernetes/helm/istio/charts/mixer/templates/config.yaml [hunk 1]
```diff
@@ -625,24 +631,27 @@ metadata:
...
   namespace: {{ .Release.Namespace }}
 spec:
-  # when running from mixer root, use the following config after adding a
-  # symbolic link to a kubernetes config file via:
-  #
-  # $ ln -s ~/.kube/config mixer/adapter/kubernetes/kubeconfig
-  #
-  # kubeconfig_path: "mixer/adapter/kubernetes/kubeconfig"
+  compiledAdapter: kubernetesenv
+  adapter: kubernetesenv
```
Comment (kyessenov) [inline install/kubernetes/helm/istio/charts/mixer/templates/config.yaml]: Per documentation, this refers to OOP adapter. `compiledAdapter` is required for in-process adapters.

Comment (douglas-reid) [reply install/kubernetes/helm/istio/charts/mixer/templates/config.yaml]: They are both listed as `Required`. I will try removing `adapter` and see what happens.

**Comment by douglas-reid** [comment]
@kyessenov any idea how concerned we should be generally about:

```
I0829 21:46:20.648] 2018-08-29T21:46:20.647523Z	error	Failed to deploy Istio.
I0829 21:46:20.648] 2018-08-29T21:46:20.647634Z	error	Failed to complete Init. Error context deadline exceeded
I0829 21:46:20.648] 2018-08-29T21:46:20.647642Z	info	Saving logs
I0829 21:46:20.649] 2018-08-29T21:46:20.647647Z	info	Creating status file
I0829 21:46:20.649] 2018-08-29T21:46:20.647971Z	info	Created Status file /home/bootstrap/go/src/istio.io/istio/_artifacts/mixer-test-7f52daf6407441e5913ce95199/mixer_test.json
I0829 21:46:20.649] 2018-08-29T21:46:20.647989Z	info	Running command kubectl get ingress --all-namespaces --kubeconfig=
I0829 21:46:20.929] 2018-08-29T21:46:20.926301Z	info	Command output: 
I0829 21:46:20.932] No resources found.
I0829 21:46:20.932] 2018-08-29T21:46:20.926358Z	info	Running command kubectl get pods -n istio-system --kubeconfig=
I0829 21:46:21.538] 2018-08-29T21:46:21.537665Z	info	Command output: 
I0829 21:46:21.538] NAME                                        READY     STATUS             RESTARTS   AGE
I0829 21:46:21.538] istio-citadel-5db665c59d-97xvm              1/1       Running            0          8m
I0829 21:46:21.538] istio-egressgateway-595d494557-vdnrs        1/1       Running            0          8m
I0829 21:46:21.539] istio-galley-74b55c948-2fvsf                0/1       CrashLoopBackOff   6          8m
I0829 21:46:21.539] istio-ingressgateway-5d74788df6-6s7cn       1/1       Running            0          8m
I0829 21:46:21.539] istio-pilot-c5b9c779b-bvw6n                 2/2       Running            0          8m
I0829 21:46:21.539] istio-policy-7b6b856548-xc7m6               2/2       Running            0          8m
I0829 21:46:21.540] istio-sidecar-injector-54bfd58f58-8q4wp     1/1       Running            0          8m
I0829 21:46:21.540] istio-statsd-prom-bridge-55965ff9c8-f5kjp   1/1       Running            0          8m
I0829 21:46:21.540] istio-telemetry-7b496d9c85-zns29            2/2       Running            0          8m
I0829 21:46:21.540] prometheus-7456f56c96-txqdz                 1/1       Running            0          8m
```
?

It seems like galley's CrashLooping is causing Istio init to fail.

**Comment by kyessenov** [comment]
@douglas-reid 

Would be good someone from galley to take a look at the crash log (@ayj, perhaps).
Not sure how this PR can trigger galley crash.

File: install/kubernetes/helm/istio/charts/mixer/templates/config.yaml (hunk 1 above)
Comment (douglas-reid) [reply install/kubernetes/helm/istio/charts/mixer/templates/config.yaml]: It all seems to work without `adapter`. Removed here and everywhere else now.

File: mixer/pkg/runtime/config/config_test.go [hunk 2]
```diff
@@ -529,6 +529,86 @@ Attributes:
...
+					Name:      "a1",
+					Namespace: "ns",
+					Kind:      "handler",
+				},
+				Type: store.Update,
+				Value: &store.Resource{
+					Spec: &configpb.Handler{
+						Name:            "a1",
+						CompiledAdapter: "adapter1",
+						Adapter:         "adapter1",
```
Comment (kyessenov) [inline mixer/pkg/runtime/config/config_test.go]: Should this be a validation error? Adapter1 is declared as a static adapter and used as an OOP adapter?

**Comment by ayj** [comment]
Galley validation is re-using the mixer validation packages. It's possible the galley integration is effected by some of the changes in this PR. Longer term we'd like to make validation more schema driven and decouple galley from mixer and pilot internals (see 
https://github.com/istio/istio/issues/4887).

Do you have a sample crash log of galley you can share (gist is fine)?

File: mixer/pkg/runtime/config/ephemeral.go [hunk 3]
```diff
@@ -220,12 +225,55 @@ func (e *Ephemeral) processAttributeManifests(ctx context.Context, errs *multier
...
+				// example: stdio.istio-system
+				Name:    key.Name + "." + key.Namespace,
+				Adapter: a,
+				Params:  a.DefaultConfig,
+			}
+			if handlerProto.Params != nil {
+				c := staticConfig.Adapter.DefaultConfig
+				switch v := handlerProto.Params.(type) {
+				case map[string]interface{}:
+					if err := convert(v, c); err != nil {
```
[the remark this answers is not in the document]
Comment (douglas-reid) [reply mixer/pkg/runtime/config/ephemeral.go]: Great catch!  Fixed.

File: mixer/pkg/runtime/config/config_test.go [hunk 4]
```diff
@@ -529,6 +529,86 @@ Attributes:
...
+  Name:    a2.ns
+  Adapter: adapter2
+  Params:  &Struct{Fields:map[string]*Value{pqr: &Value{Kind:&Value_StringValue{StringValue:abcstring,},XXX_unrecognized:[],},},XXX_unrecognized:[],}
+InstancesStatic:
+Rules:
+Attributes:
+  template.attr: BOOL
+`,
+	},
+
```
[the remark this answers is not in the document]
Comment (douglas-reid) [reply mixer/pkg/runtime/config/config_test.go]: Added two tests.

I *think* your concern may be around if they have the same name. I added a test for that. It shows dynamic adapters winning.

**Comment by douglas-reid** [comment]
@ayj for the latest e2e_simple:

```
I0830 19:17:41.008] 2018-08-30T19:17:41.007475Z	error	Failed to complete Init. Error context deadline exceeded
I0830 19:17:41.009] 2018-08-30T19:17:41.007567Z	info	Saving logs
I0830 19:17:41.009] 2018-08-30T19:17:41.007583Z	info	Creating status file
I0830 19:17:41.010] 2018-08-30T19:17:41.008000Z	info	Created Status file /home/bootstrap/go/src/istio.io/istio/_artifacts/simple-auth-test-2c3afbc7681741749ced/simple_auth_test.json
I0830 19:17:41.010] 2018-08-30T19:17:41.008018Z	info	Running command kubectl get ingress --all-namespaces --kubeconfig=
I0830 19:17:41.310] 2018-08-30T19:17:41.309768Z	info	Command output: 
I0830 19:17:41.310] No resources found.
I0830 19:17:41.310] 2018-08-30T19:17:41.309854Z	info	Running command kubectl get pods -n istio-system --kubeconfig=
I0830 19:17:41.918] 2018-08-30T19:17:41.918229Z	info	Command output: 
I0830 19:17:41.919] NAME                                        READY     STATUS             RESTARTS   AGE
I0830 19:17:41.919] echosrv-deployment-1-d946d88ff-w42qr        2/2       Running            0          8m
I0830 19:17:41.919] echosrv-deployment-2-5567b9d94b-s978s       2/2       Running            0          8m
I0830 19:17:41.919] istio-citadel-7bbb79b59b-9gd6c              1/1       Running            0          9m
I0830 19:17:41.919] istio-egressgateway-84589669f5-qrfn2        1/1       Running            0          9m
I0830 19:17:41.920] istio-galley-5d5f8594f8-dh8h9               0/1       CrashLoopBackOff   6          9m
```
-- https://storage.googleapis.com/istio-prow/pull/istio_istio/8356/e2e-simpleTests/8768/build-log.txt

With galley logs:

```
panic: unknown or mismatched type: type: interface {}, wire type: bytes

goroutine 123 [running]:
istio.io/istio/vendor/github.com/gogo/protobuf/proto.typeMarshaler(0x1df1340, 0x190b980, 0xc4209e92c0, 0x4, 0x4, 0xc4209e0000, 0x4, 0x4)
	/home/bootstrap/go/src/istio.io/istio/vendor/github.com/gogo/protobuf/proto/table_marshal.go:882 +0x1d90
istio.io/istio/vendor/github.com/gogo/protobuf/proto.(*marshalFieldInfo).setMarshaler(0xc420cadbd8, 0xc420d8b880, 0xc4209e92c0, 0x4, 0x4)
	/home/bootstrap/go/src/istio.io/istio/vendor/github.com/gogo/protobuf/proto/table_marshal.go:553 +0x9d
istio.io/istio/vendor/github.com/gogo/protobuf/proto.(*marshalFieldInfo).computeMarshalFieldInfo(0xc420cadbd8, 0xc420d8b880)
	/home/bootstrap/go/src/istio.io/istio/vendor/github.com/gogo/protobuf/proto/table_marshal.go:481 +0x1a5
istio.io/istio/vendor/github.com/gogo/protobuf/proto.(*marshalInfo).computeMarshalInfo(0xc4208cd500)
	/home/bootstrap/go/src/istio.io/istio/vendor/github.com/gogo/protobuf/proto/table_marshal.go:411 +0xb0f
istio.io/istio/vendor/github.com/gogo/protobuf/proto.(*marshalInfo).size(0xc4208cd500, 0xc420aef3b0, 0xc420605b20)
	/home/bootstrap/go/src/istio.io/istio/vendor/github.com/gogo/protobuf/proto/table_marshal.go:170 +0x3b1
istio.io/istio/vendor/github.com/gogo/protobuf/proto.(*InternalMessageInfo).Size(0xc420605b20, 0x1dba320, 0xc420aef3b0, 0x0)
	/home/bootstrap/go/src/istio.io/istio/vendor/github.com/gogo/protobuf/proto/table_marshal.go:129 +0x65
istio.io/istio/vendor/github.com/gogo/protobuf/proto.Marshal(0x1dba320, 0xc420aef3b0, 0x29da260, 0xc4209639f8, 0x40fd09, 0xc420a9cf00, 0x30)
	/home/bootstrap/go/src/istio.io/istio/vendor/github.com/gogo/protobuf/proto/table_marshal.go:2749 +0xf4
istio.io/istio/galley/pkg/runtime.(*State).envelopeResource(0xc4206d1460, 0x1, 0x1c336c7, 0x30, 0xc420054e00, 0x12, 0xc420bd074c, 0x4, 0x0, 0xed31a37b9, ...)
	/home/bootstrap/go/src/istio.io/istio/galley/pkg/runtime/state.go:133 +0x50
istio.io/istio/galley/pkg/runtime.(*State).apply(0xc4206d1460, 0x1, 0x1c336c7, 0x30, 0xc420054e00, 0x12, 0xc420bd074c, 0x4, 0x0, 0xed31a37b9, ...)
	/home/bootstrap/go/src/istio.io/istio/galley/pkg/runtime/state.go:72 +0x2bb
istio.io/istio/galley/pkg/runtime.(*Processor).processEvent(0xc420562780, 0x1, 0x1c336c7, 0x30, 0xc420054e00, 0x12, 0xc420bd074c, 0x4, 0x0, 0xed31a37b9, ...)
	/home/bootstrap/go/src/istio.io/istio/galley/pkg/runtime/processor.go:176 +0x10f
istio.io/istio/galley/pkg/runtime.(*Processor).process(0xc420562780)
	/home/bootstrap/go/src/istio.io/istio/galley/pkg/runtime/processor.go:142 +0x3cb
created by istio.io/istio/galley/pkg/runtime.(*Processor).Start
	/home/bootstrap/go/src/istio.io/istio/galley/pkg/runtime/processor.go:107 +0x22c
```

-- https://storage.googleapis.com/istio-prow/pull/istio_istio/8356/e2e-simpleTests/8768/artifacts/simple-auth-test-2c3afbc7681741749ced/istio-galley-5d5f8594f8-dh8h9_container:galley.prev.log

**Comment by douglas-reid** [comment]
@ayj @mandarjog it would appear that Galley can't handle events where the item is `"type.googleapis.com/istio.policy.v1beta1.Handler"`. Is this a known issue?

**Comment by kyessenov** [comment]
Is galley enabled by default? Why is it broken by the change in mixer CRDs?

**Comment by douglas-reid** [comment]
@kyessenov it appears that the `Params` field in `v1beta1.Handler` causes marshal issues (perhaps just when empty value). more investigation is needed. But, I think it is broken because we had no `Handler` configs in any test before.

**Comment by douglas-reid** [comment]
Ah! `Handler.params` is of type `google.protobuf.Struct`. I'm guessing that this is failure for marshaling that data.

**Comment by ayj** [comment]
```bash
$ cat handler.go 
package main

import (
        "github.com/gogo/protobuf/proto"
        cfg "istio.io/api/policy/v1beta1"
)

func main() {
        c := &cfg.Handler{Name: "foo"}
        proto.Marshal(c)
}

$ go run handler.go 
panic: unknown or mismatched type: type: interface {}, wire type: bytes

goroutine 1 [running]:
istio.io/istio/vendor/github.com/gogo/protobuf/proto.typeMarshaler(0x5db940, 0x589c00, 0xc420056080, 0x4, 0x4, 0xc420050000, 0x4, 0x4)
        /home/jasonyoung/work/src/istio.io/istio/vendor/github.com/gogo/protobuf/proto/table_marshal.go:882 +0x1d90
istio.io/istio/vendor/github.com/gogo/protobuf/proto.(*marshalFieldInfo).setMarshaler(0xc420001158, 0xc420088150, 0xc420056080, 0x4, 0x4)
        /home/jasonyoung/work/src/istio.io/istio/vendor/github.com/gogo/protobuf/proto/table_marshal.go:553 +0x9d
istio.io/istio/vendor/github.com/gogo/protobuf/proto.(*marshalFieldInfo).computeMarshalFieldInfo(0xc420001158, 0xc420088150)
        /home/jasonyoung/work/src/istio.io/istio/vendor/github.com/gogo/protobuf/proto/table_marshal.go:481 +0x1a5
istio.io/istio/vendor/github.com/gogo/protobuf/proto.(*marshalInfo).computeMarshalInfo(0xc420086000)
        /home/jasonyoung/work/src/istio.io/istio/vendor/github.com/gogo/protobuf/proto/table_marshal.go:411 +0xb0f
istio.io/istio/vendor/github.com/gogo/protobuf/proto.(*marshalInfo).size(0xc420086000, 0xc42006a140, 0xc42000a080)
        /home/jasonyoung/work/src/istio.io/istio/vendor/github.com/gogo/protobuf/proto/table_marshal.go:170 +0x3b1
istio.io/istio/vendor/github.com/gogo/protobuf/proto.(*InternalMessageInfo).Size(0xc42000a080, 0x5d99c0, 0xc42006a140, 0x0)
        /home/jasonyoung/work/src/istio.io/istio/vendor/github.com/gogo/protobuf/proto/table_marshal.go:129 +0x65
istio.io/istio/vendor/github.com/gogo/protobuf/proto.Marshal(0x5d99c0, 0xc42006a140, 0x0, 0x0, 0xc4200645a0, 0xc420045f78, 0xc420062058)
        /home/jasonyoung/work/src/istio.io/istio/vendor/github.com/gogo/protobuf/proto/table_marshal.go:2749 +0xf4
main.main()
        /home/jasonyoung/work/src/istio.io/istio/handler.go:10 +0x63
exit status 2
```

**Comment by kyessenov** [comment]
I have filed a request to use interface in place of structs in gogo/golang protos. As of now, handler config is not a valid proto and cannot be sent over the wire without using hacks into proto. We should track this in a separate issue.

**Comment by douglas-reid** [comment]
@ayj @mandarjog @kyessenov I've commented out Handler from the types info in galley in this PR. I think that might unblock this particular PR. If it does, I'll file an issue to support Handler in galley moving forward. I feel like there is some promise in all the converter stuff, but I don't know the codebase well enough to muck around in that logic here.

**Comment by ayj** [comment]
`Instance` also makes use of `interface{}` and presumably will panic on proto encoding.

**Comment by kyessenov** [comment]
Tests failing and merge conflicts...

**Comment by douglas-reid** [comment]
@kyessenov yup. updating to address. i added the clone in the wrong spot in my haste yesterday.

**Comment by kyessenov** [comment]
Please deflake circle tests.

File: galley/pkg/metadata/kube/types.go [hunk 5]
```diff
@@ -468,16 +468,16 @@ func init() {
...
-		Kind:      "handler",
-		ListKind:  "handlerList",
-		Singular:  "handler",
-		Plural:    "handlers",
-		Version:   "v1alpha2",
-		Group:     "config.istio.io",
-		Target:    metadata.Types.Get("type.googleapis.com/istio.policy.v1beta1.Handler"),
-		Converter: converter.Get("identity"),
-	})
+	// b.Add(kube.ResourceSpec{
```
Comment (kyessenov) [inline galley/pkg/metadata/kube/types.go]: @douglas-reid @ayj please file an issue about this. Galley/Mixer need to handle custom struct form in protos.

File: mixer/pkg/runtime/config/config_test.go [hunk 6]
```diff
@@ -529,6 +529,191 @@ Attributes:
...
+					Spec: &configpb.Handler{
+						Name:            "a2",
+						CompiledAdapter: "adapter2",
+						Params:          adapter2Params,
+					},
+				},
+			},
+		},
+		E: `
+ID: 0
```
Comment (kyessenov) [inline mixer/pkg/runtime/config/config_test.go]: Please let's not encourage this text-based validation. It's difficult to track references and _feels_ fragile. Why don't we use structural struct comparison?

File: mixer/pkg/runtime/config/config_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (kyessenov) [reply mixer/pkg/runtime/config/config_test.go]: Thanks.

File: mixer/pkg/server/server.go [hunk 7]
```diff
@@ -193,7 +194,13 @@ func newServer(a *Args, p *patchTable) (*Server, error) {
...
 	}
 
-	kinds := runtimeconfig.KindMap(adapterMap, templateMap)
+	var kinds map[string]proto.Message
+	if a.UseAdapterCRDs {
+		kinds = runtimeconfig.KindMap(adapterMap, templateMap)
+	} else {
+		kinds = runtimeconfig.KindMap(map[string]*adapter.Info{}, templateMap)
+	}
+
```
Comment (kyessenov) [inline mixer/pkg/server/server.go]: There is `newRuntime` call that takes adapterMap below. Do you want to change that as well?

File: galley/pkg/metadata/kube/types.go (hunk 5 above)
Comment (ayj) [reply galley/pkg/metadata/kube/types.go]: https://github.com/istio/istio/issues/8486

**Comment by mandarjog** [comment]
@douglas-reid now that Mixer's custom proto is removed, this can be picked back up?

**Comment by douglas-reid** [comment]
Yes. It just hasn't bubbled back up to the top of my priority queue :).

**Comment by mandarjog** [comment]
shell check errors are already fixed in Master.

**Pull request merged**